src/object/cli_obj.c

```diff
@@ -1430,37 +1426,60 @@ obj_iod_valid(unsigned int nr, daos_iod_t *iods, bool update)
 		switch (iods[i].iod_type) {
 		default:
 			D_ERROR("Unknown iod type=%d\n", iods[i].iod_type);
-			return false;
+			return -DER_INVAL;
 
 		case DAOS_IOD_NONE:
 			if (!iods[i].iod_recxs && iods[i].iod_nr == 0)
 				continue;
 
 			D_ERROR("IOD_NONE ignores value iod_nr=%d, recx=%p\n",
 				 iods[i].iod_nr, iods[i].iod_recxs);
-			return false;
+			return -DER_INVAL;
 
 		case DAOS_IOD_ARRAY:
 			if (!update && iods[i].iod_nr == 0) /* size query */
 				continue;
 
-			if (obj_recx_valid(iods[i].iod_nr, iods[i].iod_recxs,
+			if (sgls == NULL) {
+				/* size query or punch */
+				if (!update || iods[i].iod_size == 0)
+					continue;
```

Comment:
hmm, this check is really confused. I would prefer you use opc to skip, instead by update && and iod_size.

and also why check the size query again?  I thought the above (!update && iods[i].iod_nr == 0) is for size query?

Reply:
obj_iod_sgl_valid() is only called for update or fetch, so it did not use opc to check it.
size query can with iod_nr = 0, and NULL sgl,  and punch record (see test in punch_rec_with_rxnr) can with null sgl and iod_size = 0.
I feel a little confuse as well, but without it cannot pass daos_test -i.

Reply:
case ARRAY:
         if (iods[i].iod_size == 0)   /* size query or punch, in this case even sgls != NULL, but since iod_size = 0, so it can not check if sgls can match with iods any way */
                    continue;

        obj_recx_valid();
        if (sgls != NULL) {
                /* check whether sgls match with iods */
                obj_req_size_valid();
        }

Reply:
for normal update, better to check sgl's valid that cannot be checked by above suggested code.

Reply:
hmm, normal update, iod_size should not be 0, otherwise it is punch. or I miss sth here?

Reply:
I mean for if a update take iod_size != 0 but with NULL sgls, that is invalid but cannot catch by you posted code.

Reply:
ah, you mean this?
      ```
       if (sgls != NULL) {
                obj_req_size_valid();
       } else {
               if (update)
                       return -DER_INVAL;
       }
```